### ISSUE TITLE
Javascript Tooling

### ISSUE THREAD

**Issue by hzoo** [syg/ecmascript-binary-ast#11]
TL;DR - this is cool because it unifies the JS ecosystem AST, and can lead to perf when compiling JS in the current toolchain.

If I or someone else doesn't bring it up at the meeting, I'l just note here that another reason that this proposal is a 👍  from me is that it would help to unify the AST of the javascript ecosystem of tools (especially regarding babel, webpack, uglify, other parsers like acorn, esprima, etc). https://github.com/estree/estree was started as most know to unify this effort but it's diverted in certain ways (babel 5 to babel 6 https://github.com/babel/babylon/#output although we have a compat plugin).

I believe some people wanted to do an ESTree 2 (I believe the reasoning that Sebastian changed the AST in Babel 6 was just that the old AST was difficult to use when trying to create Babel plugins). If this becomes part of JS then it could be a nice forcing function for the tools that need to output the binary AST and everyone else that opts in to this could give perf benefits to developers given we can pass the same AST between all tools (ex: babel -> webpack -> uglify) and not have to re-parse in that toolchain as well as in the browser itself.

Discussion on the actual AST is a bit different but ya just saying the idea is enticing!

**Comment by Yoric** [comment]
I was also thinking about the comments. If we don't want comments to affect parse time, it will complicate a bit what we have in mind for tokenization, but that shouldn't be too hard.

**Comment by hzoo** [comment]
Parsing comments do make babylon slower atm (and technically maybe the browser doesn't need to retain them but at least the tooling in between that uses the AST may/should need to)

**Comment by Yoric** [comment]
Quick summary of my latest idea: store comments at the end of the file, after the AST. This shouldn't complicate things.

**Comment by hzoo** [comment]
Yeah babylon/most parsers do this already (comments property) http://astexplorer.net/#/gist/c3dd0d6473d7eb8fdf542aac21ab595f/20bca45ef3d0ce3f59c4cf371d4977ebd5a22cff, although babylon adds leadingComments/trailingComments for nodes

**Comment by mikesherov** [comment]
One thing to keep in mind is that several tools have significance in comments: `eslint` has them to disable: `// eslint-disable`, Webpack has them to indicate chunkName of imports: `import( /* webpackChunkName: whatever */ 'file.js');`, and so associating the comments with a particular node (either via inert annotations in the new AST format or via comments directly)  would be pretty important to preserve.
